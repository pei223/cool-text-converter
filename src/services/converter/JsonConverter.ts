import { infoLog } from "../../utils/logger";
import { ConvertibleMarkdownFormats } from "../consts/format";
import { MarkdownUnrecognizedError } from "../errors/types";
import { writeMarkdownTable } from "./tableWriter";
import { MarkdownConverter } from "./types";

export class JsonConverter implements MarkdownConverter {
  recognize(v: string): void {
    let parsed: unknown;
    try {
      parsed = JSON.parse(v);
    } catch (e) {
      throw new MarkdownUnrecognizedError(v, "JSON", `Not parsed: ${e}`);
    }
    // typeof で配列はobjectを返す.
    // https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/typeof#%E5%9F%BA%E6%9C%AC%E7%9A%84%E3%81%AA%E4%BD%BF%E3%81%84%E6%96%B9
    // nullもtypeofだとobject判定になる.
    // https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/typeof#typeof_null
    if (typeof parsed !== "object" || Array.isArray(parsed) || parsed == null) {
      throw new MarkdownUnrecognizedError(
        v,
        "JSON",
        `Parsed, but not object: ${parsed}`
      );
    }
  }
  getDetailedConverter(): MarkdownConverter | null {
    return null;
  }
  convertToMarkdown(v: string): string {
    try {
      const parsed = JSON.parse(v);
      return writeMarkdownTable(["key", "value"], Object.entries(parsed));
    } catch (e) {
      // 基本的にrecognizeに成功しているか、ユーザーが明示的に指定しないと起きない.
      infoLog("Parse error on JSON converteToMarkdown", e);
      return "JSON形式として認識できませんでした";
    }
  }
  formatName(): ConvertibleMarkdownFormats {
    return "JSON";
  }
  formatLabel(): string {
    return "JSON";
  }
}
