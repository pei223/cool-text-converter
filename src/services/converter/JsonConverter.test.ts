import { describe, it, expect } from "vitest";
import { JsonConverter } from "./JsonConverter";

describe("URLConverter", () => {
  const converter = new JsonConverter();

  describe("recognize", () => {
    it.each([
      '{"test": 1}',
      '{"test": "1"}',
      '{"test": "1", "hoge": 2}',
      '{"child": {"hoge": 1}}',
      '{"child1": {"child2": {"hoge": 1}}}',
      '{"test": [1, 2]}',
    ])("Success(%s)", (v: string) => {
      expect(() => converter.recognize(v)).not.toThrow();
    });
    it.each(["http//hoge", "1234", "invalid", "[1,2,3]", '["1", "2"]'])(
      "Not apply(%s)",
      (v: string) => {
        expect(() => converter.recognize(v)).toThrow();
      }
    );
  });
  // 中身いちいち正解かくのしんどいのでスナップショットテストにする
  it.each([
    ['{"test": 1}'],
    ['{"test": "1"}'],
    ['{"test": "1", "hoge": 2}'],
    ['{"child": {"hoge": 1}}'],
    ['{"child1": {"child2": {"hoge": 1}}}'],
    ['{"test": [1, 2]}'],
  ])("converteToMarkdown(%s) - (%s)", (v: string) => {
    expect(converter.convertToMarkdown(v)).toMatchSnapshot();
  });
});
