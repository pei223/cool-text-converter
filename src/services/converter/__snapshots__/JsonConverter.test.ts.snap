// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JsonConverter > converteToMarkdown({"child": {"hoge": 1}}) - (%s) 1`] = `
"| key | value |
| ---- | ---- |
| child | {"hoge":1} |"
`;

exports[`JsonConverter > converteToMarkdown({"child1": {"child2": {"hoge": 1}}}) - (%s) 1`] = `
"| key | value |
| ---- | ---- |
| child1 | {"child2":{"hoge":1}} |"
`;

exports[`JsonConverter > converteToMarkdown({"test": "1", "hoge": 2}) - (%s) 1`] = `
"| key | value |
| ---- | ---- |
| test | 1 |
| hoge | 2 |"
`;

exports[`JsonConverter > converteToMarkdown({"test": "1"}) - (%s) 1`] = `
"| key | value |
| ---- | ---- |
| test | 1 |"
`;

exports[`JsonConverter > converteToMarkdown({"test": [1, 2]}) - (%s) 1`] = `
"| key | value |
| ---- | ---- |
| test | 1,2 |"
`;

exports[`JsonConverter > converteToMarkdown({"test": 1}) - (%s) 1`] = `
"| key | value |
| ---- | ---- |
| test | 1 |"
`;
